craft_parts/plugins/gradle_plugin.py

```diff
@@ -0,0 +1,165 @@
+        return self
+
+
+class GradlePluginEnvironmentValidator(validator.PluginEnvironmentValidator):
+    """Check the execution environment for the gradle plugin."""
+
+    @override
+    def validate_environment(
+        self, *, part_dependencies: list[str] | None = None
+    ) -> None:
+        """Ensure the environment contains dependencies needed by the plugin.
+
+        :param part_dependencies: A list of the parts this part depends on.
+        :raises PluginEnvironmentValidationError: If gradle is invalid
+          and there are no parts named gradle-deps.
+        """
+        # We can't check for gradle executable here because if the project uses gradlew, we don't
+        # necessarily need gradle to be installed on the system. We can't check if the project uses
+        # gradlew since the project would not be pulled yet.
+
+
+class GradlePlugin(JavaPlugin):
+    """A plugin to build parts that use Gradle.
+
+    The Gradle build system is commonly used to build Java projects. This
+    plugin requires a pom.xml in the root of the source tree.
+    This plugin uses the common plugin keywords as well as those for "sources".
+    For more information check the 'plugins' topic for the former and the
+    'sources' topic for the latter.
+    Additionally, this plugin uses the following plugin-specific keywords:
+    - gradle-parameters:
+      (list of strings)
+      Flags to pass to the build using the gradle semantics for parameters.
+    """
+
+    properties_class = GradlePluginProperties
+    validator_class = GradlePluginEnvironmentValidator
+
+    @property
+    def gradle_executable(self) -> str:
+        """Use gradlew by default if it exists."""
+        return (
+            f"{self._part_info.part_build_dir}/gradlew"
+            if (self._part_info.part_build_dir / "gradlew").exists()
+            else "gradle"
+        )
+
+    @override
+    def get_build_snaps(self) -> set[str]:
+        """Return a set of required snaps to install in the build environment."""
+        return set()
+
+    @override
+    def get_build_packages(self) -> set[str]:
+        """Return a set of required packages to install in the build environment."""
+        return set()
+
+    @override
+    def get_build_commands(self) -> list[str]:
+        """Return a list of commands to run during the build step."""
+        options = cast(GradlePluginProperties, self._options)
+
+        gradle_cmd = [self.gradle_executable, options.gradle_task]
+        self._setup_proxy()
+
+        return [
+            *self._get_gradle_init_command(options=options),
+            " ".join(gradle_cmd + options.gradle_parameters),
+            # remove gradle-wrapper.jar files included in the project if any.
+            f'find {self._part_info.part_build_dir} -name "gradle-wrapper.jar" -type f -delete',
+            *self._get_java_post_build_commands(),
+        ]
+
+    def _setup_proxy(self) -> None:
```

Comment:
It looks like some of this is similar to maven, do you think the logic can be unified for that part?

Reply:
That is a good point - but unfortunately at the moment, it might be better to wait for either of them to get merged first so that a merge conflict would be avoided and then we can refactor the common functions together :)

Reply:
ok makes sense